### DISCUSSION TITLE
Task.ContinueWith OnlyOnRanToCompletion Shortcut: ->

### DISCUSSION THREAD

**Discussion by pavlindrom** [dotnet/csharplang#7132] [Language Ideas]
If you squint really hard, and look from the corner of your eye, you can notice some similarities between *Task*s and *pointer*s.

They are both references to an entity in order to optimize resources. Tasks reference a future value to optimize CPU usage, while pointers reference a memory location to use less RAM.

Trying to access the result requires dereferencing on both, using `await thing` or `*thing`, and when trying to access a member of the result needs surrounding with parentheses.

Given that connection, I think it would make sense for Tasks to borrow an operator from pointers: `->`

This will turn:

```csharp
var midResult1 = await myVar.OpOneAsync();
var andAfterOp2 = await midResult1.OpTwoAsync();
var andMore = andAfterOp2.SyncOpThree();
return await andMore.OpFourAsync();
```

into:

```csharp
return await myVar.OpOneAsync()
    ->OpTwoAsync()
    ->SyncOpThree()
    ->OpFourAsync();
```

We could think of `->` as a call to `Task.ContinueWith`:

```csharp
await myVar.OpOneAsync()
    .ContinueWith(async t => await t.Result.OpTwoAsync(),
        TaskContinuationOptions.OnlyOnRanToCompletion)
    .ContinueWith(t => t.Unwrap().Result.SyncOpThree(),
        TaskContinuationOptions.OnlyOnRanToCompletion)
    .ContinueWith(async t => await t.Result.OpFourAsync(),
        TaskContinuationOptions.OnlyOnRanToCompletion);
```

The usefulness in thinking of the implementation as though you're calling ContinueWith is that it allows for mixing async and non-async code seamlessly.

I think this feature will allow some fluent syntax for async code, which can improve code readability.

**Comment by HaloFour** [comment]
> I think this feature will allow some fluent syntax for async code, which can improve code readability.

I'd like to see a way to fluently chain asynchronous code, but I disagree that conflating tasks with pointers is the way to achieve that.

**Comment by TahirAhmadov** [comment]
I think what you're looking for is the "pipe forward await" feature, which has been requested and discussed, but I can't find a specific discussion for it at the moment - possible I saw comments about it under other discussions, such as [this one](https://github.com/dotnet/csharplang/discussions/74#discussioncomment-64431).

**Comment by pavlindrom** [comment]
Well no, I thought this would make most sense for C# since it's based on C++.  I think arrow makes sense and is clear about what it does. Still, it's up to the community and the designers to see if it fits.

**Comment by huoyaoyuan** [comment]
It can be achieved by a "postfix await" operator. See also #96.

**Comment by En3Tho** [comment]
Seeing how tasks and pointers are so similar I also propose awaiting pointers to to get AV's reliably and also chancing adressof operator on tasks to return a pointer to the future.
